ext/protos_test.go

```diff
@@ -0,0 +1,159 @@
+// limitations under the License.
+
+package ext
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/common/types/ref"
+	"github.com/google/cel-go/common/types/traits"
+	"github.com/google/cel-go/test"
+	"github.com/google/cel-go/test/proto2pb"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/wrapperspb"
+)
+
+func TestProtos(t *testing.T) {
+	var protosTests = []struct {
+		expr      string
+		err       string
+		parseErr  string
+		parseOnly bool
+	}{
+		// Positive test cases.
+		{expr: `proto.getExt(ExampleType{}, google.expr.proto2.test.int32_ext) == 0`},
+		{expr: `!proto.hasExt(ExampleType{}, google.expr.proto2.test.int32_wrapper_ext)`},
+		{expr: `proto.getExt(ExampleType{}, google.expr.proto2.test.int32_wrapper_ext) == null`},
+		{expr: `!proto.hasExt(ExampleType{}, google.expr.proto2.test.nested_example)`},
+		{expr: `proto.getExt(ExampleType{}, google.expr.proto2.test.nested_example) == ExampleType{}`},
+		{expr: `proto.getExt(ExampleType{}, google.expr.proto2.test.ExtendedExampleType.extended_examples) == []`},
+		{expr: `proto.getExt(ExampleType{}, google.expr.proto2.test.ExtendedExampleType.enum_ext) == GlobalEnum.GOO`},
+		{expr: `proto.getExt(msg, google.expr.proto2.test.int32_ext) == 42`},
+		{expr: `proto.getExt(msg, google.expr.proto2.test.int32_wrapper_ext) == 21`},
+		{expr: `proto.hasExt(msg, google.expr.proto2.test.nested_example)`},
+		{expr: `proto.getExt(msg, google.expr.proto2.test.nested_example) == ExampleType{name: 'nested'}`},
+		{expr: `proto.getExt(msg, google.expr.proto2.test.ExtendedExampleType.extended_examples) == ['example1', 'example2']`},
+		{expr: `proto.getExt(msg, google.expr.proto2.test.ExtendedExampleType.enum_ext) == GlobalEnum.GAZ`},
+
+		// Negative test cases.
+		{
+			expr:      `proto.getExt(ExtendedExampleType{}, google.expr.proto2.test.ExtendedExampleType.enum_ext)`,
+			err:       "no such field 'google.expr.proto2.test.ExtendedExampleType.enum_ext'",
+			parseOnly: true,
+		},
+		{
+			expr: `proto.getExt(ExtendedExampleType{}, enum_ext)`,
+			parseErr: `ERROR: <input>:1:37: invalid extension field
+			| proto.getExt(ExtendedExampleType{}, enum_ext)
+			| ....................................^`,
+		},
+		{
+			expr: `proto.hasExt(ExtendedExampleType{}, call().enum_ext)`,
+			parseErr: `ERROR: <input>:1:43: invalid extension field
+			| proto.hasExt(ExtendedExampleType{}, call().enum_ext)
+			| ..........................................^`,
+		},
+		{
+			expr: `proto.getExt(ExtendedExampleType{}, has(google.expr.proto2.test.int32_ext))`,
+			parseErr: `ERROR: <input>:1:40: invalid extension field
+			| proto.getExt(ExtendedExampleType{}, has(google.expr.proto2.test.int32_ext))
+			| .......................................^`,
+		},
+
+		// Even though 'getExt' is the macro, the call is left unexpanded is an identifier
+		// that's not named 'proto'
+		{
+			expr: `msg.getExt("google.expr.proto2.test.int32_ext", 0) == 42`,
```

Comment:
Could we have a similar test for `msg.hasExt` as well

Reply:
Done.